gate-web/src/main/groovy/com/netflix/spinnaker/gate/ratelimit/scoring/ScoringRedisRateLimiter.java

```diff
@@ -0,0 +1,186 @@
+
+      // Get the remaining capacity for the principal
+      final long now = clock.millis();
+      final long window = clock.instant().minus(Duration.ofHours(1)).toEpochMilli();
+
+      // Age-out old requests, range the set of requests for the principal and reset the expiry
+      Transaction txn = jedis.multi();
+      txn.zremrangeByScore(key, 0, window);
+      txn.zrange(key, 0, -1);
+      txn.expire(key, (int) window / 1_000);
+      List<Object> result = txn.exec();
+
+      // zrange result: History of previously used request in the window
+      @SuppressWarnings("unchecked")
+      final int usedRequests =
+          ((Set<String>) result.get(1)).stream().map(Integer::valueOf).reduce(0, Integer::sum);
+
+      if (usedRequests == 0) {
+        // It's cool dog, just allow it regardless of cost or capacity.
+        rate.setThrottled(false);
+        rate.setRemaining(principal.getCapacity() - score);
+      } else if (usedRequests + score > principal.getCapacity()) {
+        rate.setThrottled(true);
+        rate.setRemaining(principal.getCapacity());
+      } else {
+        rate.setThrottled(false);
+        rate.setRemaining(principal.getCapacity() - (usedRequests + score));
+      }
+
+      // Add the request if this request is actually accepted
+      if (!rate.isThrottled()) {
+        jedis.zadd(key, now, String.valueOf(score));
+      }
+    }
+
+    rate.setRateSeconds(1);
+    rate.setCapacity(principal.getCapacity());
+    rate.setReset(Instant.now().plus(Duration.ofHours(1)).toEpochMilli());
+
+    return rate;
+  }
+
+  @Override
+  public int calculateTotalCost(
+      @Nonnull HttpServletRequest request,
+      @Nonnull HttpServletResponse response,
+      @Nullable String handlerMethod) {
+    final int score = scoreJudge.score(request, response);
+
+    if (handlerMethod == null) {
+      // If the handlerMethod couldn't be resolved, we can't save this invocation, so just return
+      // the score.
+      return score;
+    }
+
+    // Persist the actual total score of the request, keeping the last 100 scores for the handler.
+    final String key = handlerKey(handlerMethod);
+    try (Jedis jedis = jedisPool.getResource()) {
+      Pipeline p = jedis.pipelined();
+      p.rpush(key, String.valueOf(score));
+      p.ltrim(key, 0, NUM_SCORES - 1);
+      p.expire(key, 3_600);
+      p.sync();
+    } catch (Exception e) {
+      log.error("Failed updating request score ({}) for '{}'", score, key, e);
+      return score;
+    }
+
+    return score;
+  }
+
+  private int getRequestScore(HttpServletRequest request, String handlerMethod) {
+    if (handlerMethod == null) {
+      // No handlerMethod, so we can only operate off of whatever score is generated from the
+      // request itself.
+      return scoreJudge.score(request, null);
+    }
+
+    final String key = handlerKey(handlerMethod);
+    try (Jedis jedis = jedisPool.getResource()) {
+      return scoreJudge.historicalScore(
```

Comment:
I don't understand this historical score business, it's not referring to the principal at all?

Reply:
Correct. This is the historical cost of the request handler, which the principal has no bearing on. If someone tries to load an application (say, `GET /applications/gate`) the cost of this request is going to be roughly the same if you or I request it.